File: src/tinymonitor/monitor.py
import time
import argparse
import logging
from tinymonitor import __version__
from tinymonitor.config import load_config
from tinymonitor.alerts.manager import AlertManager
from tinymonitor.metrics import CpuMetric, MemoryMetric, FilesystemMetric, LoadMetric, RebootMetric, IoMetric

logger = logging.getLogger("tinymonitor")

class Monitor:
    def __init__(self, config):
        self.config = config
        self.alert_manager = AlertManager(config)
        self.last_alert = {} # Flat dict: component_name -> timestamp
        self.alert_states = {}
        self.metrics = []
        self.load_metrics()

    def load_metrics(self):
        # Load standard metrics
        # In the future, this could be dynamic
        if self.config.get("cpu", {}).get("enabled", True):
            self.metrics.append(CpuMetric("cpu", self.config.get("cpu", {})))
        
        if self.config.get("memory", {}).get("enabled", True):
            self.metrics.append(MemoryMetric("memory", self.config.get("memory", {})))
            
        if self.config.get("filesystem", {}).get("enabled", True):
            self.metrics.append(FilesystemMetric("filesystem", self.config.get("filesystem", {})))
            
        if self.config.get("load", {}).get("enabled", True):
            self.metrics.append(LoadMetric("load", self.config.get("load", {})))

        if self.config.get("reboot", {}).get("enabled", True):
            self.metrics.append(RebootMetric("reboot", self.config.get("reboot", {})))

        if self.config.get("io", {}).get("enabled", True):
            self.metrics.append(IoMetric("io", self.config.get("io", {})))

    def process_state(self, component, level, value, duration):
        """
        Manages alert state persistence.
        Returns True if the alert should be triggered, False otherwise.
        """
        if level is None:
            # Return to normal: clear state
            if component in self.alert_states:
                # Only log recovery if we actually triggered an alert state (duration passed)
                if self.alert_states[component].get("alert_triggered", False):
                    logger.info(f"RECOVERY: {component} is back to normal")
                del self.alert_states[component]
            return False

        now = time.time()
        current_state = self.alert_states.get(component)
        
        if not current_state or current_state["level"] != level:
            self.alert_states[component] = {
                "level": level,
                "start_time": now,
                "alert_triggered": False
            }
            if duration <= 0:
                self.alert_states[component]["alert_triggered"] = True
                return True
            else:
                logger.debug(f"Detected {component} {level}, waiting for duration {duration}s")
                return False
        
        elapsed = now - current_state["start_time"]
        if elapsed >= duration:
            self.alert_states[component]["alert_triggered"] = True
            return True
            
        return False

    def trigger_alert(self, component, level, value):
        current_time = time.time()
        last_time = self.last_alert.get(component, 0)

        cooldown = self.config.get("cooldown", 60)
        should_alert = False

        if cooldown < 0:
            # "Alert Once" Mode
            state = self.alert_states.get(component)
            if state:
                start_time = state["start_time"]
                if last_time < start_time:
                    should_alert = True
        else:
            # Classic Mode
            if (current_time - last_time) > cooldown:
                should_alert = True

        if should_alert:
            logger.info(f"ALERT: {component} is {level} ({value})")
            self.alert_manager.send_alert(component, level, value)
            self.last_alert[component] = current_time
        else:
            if cooldown >= 0:
                logger.debug(f"Alert suppressed (cooldown): {component}")
            else:
                logger.debug(f"Alert suppressed (already sent): {component}")

    def run(self):
        logger.info("Starting TinyMonitor...")
        try:
            while True:
                for metric in self.metrics:
                    try:
                        results = metric.check()
                        for component, level, value in results:
                            duration = metric.config.get("duration", 0)
                            if self.process_state(component, level, value, duration):
                                self.trigger_alert(component, level, value)
                    except Exception as e:
                        logger.error(f"Error checking metric {metric.name}: {e}")
                
                time.sleep(self.config.get("refresh", 5))
        except KeyboardInterrupt:
            try:
                logger.info("Stopping TinyMonitor...")
            except KeyboardInterrupt:
                pass

def main():
    parser = argparse.ArgumentParser(description="TinyMonitor System Monitor")
    parser.add_argument("-c", "--config", help="Path to configuration file")
    parser.add_argument("-v", "--version", action="version", version=f"TinyMonitor {__version__}")
    args = parser.parse_args()

    config = load_config(args.config)

    # Configure logging
    handlers = [logging.StreamHandler()]
    log_file = config.get("log_file")
    if log_file:
        handlers.append(logging.FileHandler(log_file))

    logging.basicConfig(
        level=logging.INFO,
        format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
        handlers=handlers,
        force=True
    )

    monitor = Monitor(config)
    try:
        monitor.run()
    except KeyboardInterrupt:
        pass

if __name__ == "__main__":
    main()

File: internal/monitor/monitor.go
package monitor

import (
	"context"
	"log/slog"
	"time"

	"github.com/Gu1llaum-3/tinymonitor/internal/alerts"
	"github.com/Gu1llaum-3/tinymonitor/internal/config"
	"github.com/Gu1llaum-3/tinymonitor/internal/metrics"
	"github.com/Gu1llaum-3/tinymonitor/internal/models"
)

// StateChange represents the result of processing a metric state
type StateChange struct {
	ShouldAlert   bool
	IsRecovery    bool
	Level         models.Severity
	PreviousLevel models.Severity
}

// Monitor is the main monitoring loop
type Monitor struct {
	config       *config.Config
	alertManager *alerts.Manager
	collectors   []metrics.Collector
	lastAlert    map[string]time.Time
	alertStates  map[string]*models.AlertState
}

// New creates a new Monitor
func New(cfg *config.Config) *Monitor {
	m := &Monitor{
		config:       cfg,
		alertManager: alerts.NewManager(cfg.Alerts),
		collectors:   make([]metrics.Collector, 0),
		lastAlert:    make(map[string]time.Time),
		alertStates:  make(map[string]*models.AlertState),
	}

	m.loadCollectors()
	return m
}

func (m *Monitor) loadCollectors() {
	if m.config.CPU.Enabled {
		m.collectors = append(m.collectors, metrics.NewCPUCollector(m.config.CPU))
	}

	if m.config.Memory.Enabled {
		m.collectors = append(m.collectors, metrics.NewMemoryCollector(m.config.Memory))
	}

	if m.config.Filesystem.Enabled {
		m.collectors = append(m.collectors, metrics.NewDiskCollector(m.config.Filesystem))
	}

	if m.config.Load.Enabled {
		m.collectors = append(m.collectors, metrics.NewLoadCollector(m.config.Load))
	}

	if m.config.Reboot.Enabled {
		m.collectors = append(m.collectors, metrics.NewRebootCollector(m.config.Reboot))
	}

	if m.config.IO.Enabled {
		m.collectors = append(m.collectors, metrics.NewIOCollector(m.config.IO))
	}
}

// processState manages alert state persistence
// Returns StateChange with information about what action to take
func (m *Monitor) processState(component string, level *models.Severity, value string, duration int) StateChange {
	if level == nil {
		// Return to normal: check if we need to send recovery
		if state, exists := m.alertStates[component]; exists {
			if state.AlertTriggered {
				previousLevel := state.Level
				delete(m.alertStates, component)
				return StateChange{
					ShouldAlert:   true,
					IsRecovery:    true,
					PreviousLevel: previousLevel,
				}
			}
			delete(m.alertStates, component)
		}
		return StateChange{ShouldAlert: false}
	}

	now := time.Now()
	currentState := m.alertStates[component]

	if currentState == nil || currentState.Level != *level {
		// Preserve AlertTriggered if transitioning to a lower severity
		preserveTriggered := false
		if currentState != nil && currentState.AlertTriggered {
			// Check if we're going to a lower severity
			if isLowerSeverity(*level, currentState.Level) {
				preserveTriggered = true
			}
		}

		m.alertStates[component] = &models.AlertState{
			Level:          *level,
			StartTime:      now,
			AlertTriggered: preserveTriggered,
		}

		// If duration is 0 OR we preserved an already-triggered alert, send immediately
		if duration <= 0 || preserveTriggered {
			m.alertStates[component].AlertTriggered = true
			return StateChange{
				ShouldAlert: true,
				IsRecovery:  false,
				Level:       *level,
			}
		}

		slog.Debug("Detected alert, waiting for duration",
			"component", component,
			"level", *level,
			"duration", duration)
		return StateChange{ShouldAlert: false}
	}

	elapsed := now.Sub(currentState.StartTime)
	if elapsed >= time.Duration(duration)*time.Second && !currentState.AlertTriggered {
		m.alertStates[component].AlertTriggered = true
		return StateChange{
			ShouldAlert: true,
			IsRecovery:  false,
			Level:       *level,
		}
	}

	return StateChange{ShouldAlert: false}
}

// isLowerSeverity returns true if newLevel is less severe than oldLevel
func isLowerSeverity(newLevel, oldLevel models.Severity) bool {
	severityOrder := map[models.Severity]int{
		models.SeverityWarning:  1,
		models.SeverityCritical: 2,
	}
	return severityOrder[newLevel] < severityOrder[oldLevel]
}

// triggerAlert sends an alert with rate limiting
func (m *Monitor) triggerAlert(component string, level models.Severity, value string) {
	currentTime := time.Now()
	lastTime := m.lastAlert[component]

	cooldown := m.config.Cooldown
	shouldAlert := false

	if cooldown < 0 {
		// "Alert Once" Mode
		state := m.alertStates[component]
		if state != nil {
			if lastTime.Before(state.StartTime) {
				shouldAlert = true
			}
		}
	} else {
		// Classic Mode
		if currentTime.Sub(lastTime) > time.Duration(cooldown)*time.Second {
			shouldAlert = true
		}
	}

	if shouldAlert {
		slog.Info("ALERT",
			"component", component,
			"level", level,
			"value", value)
		m.alertManager.SendAlert(component, level, value)
		m.lastAlert[component] = currentTime
	} else {
		if cooldown >= 0 {
			slog.Debug("Alert suppressed (cooldown)", "component", component)
		} else {
			slog.Debug("Alert suppressed (already sent)", "component", component)
		}
	}
}

// triggerRecovery sends a recovery notification (no cooldown)
func (m *Monitor) triggerRecovery(component string, previousLevel models.Severity, value string) {
	if !m.config.Alerts.SendRecovery {
		slog.Debug("Recovery notification disabled", "component", component)
		return
	}

	slog.Info("RECOVERY",
		"component", component,
		"previous_level", previousLevel,
		"value", value)
	m.alertManager.SendRecovery(component, previousLevel, value)

	// Clear the lastAlert time so future alerts aren't affected
	delete(m.lastAlert, component)
}

// Run starts the monitoring loop
func (m *Monitor) Run(ctx context.Context) {
	slog.Info("Starting TinyMonitor...")

	ticker := time.NewTicker(time.Duration(m.config.Refresh) * time.Second)
	defer ticker.Stop()

	// Run initial check immediately
	m.runChecks()

	for {
		select {
		case <-ctx.Done():
			slog.Info("Stopping TinyMonitor...")
			m.alertManager.Shutdown()
			return
		case <-ticker.C:
			m.runChecks()
		}
	}
}

func (m *Monitor) runChecks() {
	for _, collector := range m.collectors {
		results := collector.Check()
		for _, result := range results {
			duration := collector.Duration()
			change := m.processState(result.Component, result.Level, result.Value, duration)

			if change.ShouldAlert {
				if change.IsRecovery {
					m.triggerRecovery(result.Component, change.PreviousLevel, result.Value)
				} else {
					m.triggerAlert(result.Component, change.Level, result.Value)
				}
			}
		}
	}
}
